File: api/app/schemas/attendance.py
from pydantic import BaseModel
from datetime import datetime

from .lesson_instance import LessonInstanceCompactOut
from .student import StudentOut


class AttendanceIn(BaseModel):
    student_id: int
    lesson_id: int
    status: str
    manual_adjustment: bool

class AttendanceOut(BaseModel):
    id: int
    student: StudentOut
    lesson: LessonInstanceCompactOut
    status: str
    manual_adjustment: bool
    created_at: datetime

class IAttendanceRaw(BaseModel):
    id: int
    student_id: int
    lesson_id: int
    status: str
    manual_adjustment: bool
    created_at: datetime

class AttendanceCompact(BaseModel):
    id: int
    student: StudentOut
    lesson_id: int
    status: str
    manual_adjustment: bool
    created_at: datetime

class StudentAttendanceStatsOut(BaseModel):
    subject: str
    present_percent: float
    present_count: int
    absent_count: int
    late_count: int
    total: int

File: mobile/types/Attendance.ts
import {IStudent} from "@/types/Student";
import {ILessonInstance} from "@/types/schedule";

export type IAttendanceStatus = "present" | "late" | "absent";

export interface IAttendance {
    id: number;
    student: IStudent;
    lesson: ILessonInstance;
    status: IAttendanceStatus;
    manual_adjustment: boolean;
    created_at: Date;
}
export interface IAttendanceRaw {
    id: number;
    student_id: number;
    lesson_id: number;
    status: IAttendanceStatus;
    manual_adjustment: boolean;
    created_at: Date;
}
export interface IAttendanceIn {
    student_id: number;
    lesson_id: number;
    status: IAttendanceStatus;
    manual_adjustment: boolean;
}

export type IAttendanceStats = IAttendanceStatsItem[];

export interface IAttendanceStatsItem {
    subject: string;
    present_percent: number;
    present_count: number;
    total: number;
    absent_count: number;
    late_count: number;
}
